hie-plugin-api/Haskell/Ide/Engine/PluginsIdeMonads.hs

```diff
@@ -220,12 +227,57 @@ instance ToJSON IdePlugins where
 -- | IdeM that allows for interaction with the ghc-mod session
 type IdeGhcM = GM.GhcModT IdeM
 
+-- | Run an IdeGhcM with Cradle found from the current directory
+runIdeGhcM :: GM.Options -> Maybe (Core.LspFuncs Config) -> TVar IdeState -> IdeGhcM a -> IO a
+runIdeGhcM ghcModOptions mlf stateVar f = do
+  let env = IdeEnv mlf
+  (eres, _) <- flip runReaderT stateVar $ flip runReaderT env $ GM.runGhcModT ghcModOptions f
+  case eres of
+      Left err  -> liftIO $ throwIO err
+      Right res -> return res
+
 -- | A computation that is deferred until the module is cached.
 -- Note that the module may not typecheck, in which case 'UriCacheFailed' is passed
 data Defer a = Defer FilePath (UriCacheResult -> a) deriving Functor
 type IdeDeferM = FreeT Defer IdeM
 
-type IdeM = ReaderT ClientCapabilities (MultiThreadState IdeState)
+type IdeM = ReaderT IdeEnv (MultiThreadState IdeState)
+
+-- | Run an IdeM 
+runIdeM :: Maybe (Core.LspFuncs Config) -> TVar IdeState -> IdeM a -> IO a
+runIdeM mlf stateVar f = do
+  let env = IdeEnv mlf
+  flip runReaderT stateVar $ runReaderT f env
+
+getClientCapabilities :: IdeM ClientCapabilities
+getClientCapabilities = do
+  IdeEnv mlf <- ask
+  case mlf of
+    Just lf -> return (Core.clientCapabilities lf)
+    Nothing -> return def
+
+getConfig :: IdeM Config
+getConfig = do
+  IdeEnv mlf <- ask
+  case mlf of
+    Just lf -> fromMaybe def <$> liftIO (Core.config lf)
+    Nothing -> return def
+
+getVirtualFile :: Uri -> IdeM (Maybe VirtualFile)
+getVirtualFile uri = do
+  IdeEnv mlf <- ask
+  case mlf of
+    Just lf -> liftIO $ Core.getVirtualFileFunc lf uri
+    Nothing -> return Nothing
+
+getRootPath :: IdeM (Maybe FilePath)
+getRootPath = do
+  IdeEnv mlf <- ask
+  case mlf of
+    Just lf -> return (Core.rootPath lf)
+    Nothing -> return Nothing
+
+data IdeEnv = IdeEnv (Maybe (Core.LspFuncs Config))
```

Comment:
Can you make this a `newtype` instead?

Reply:
I meant to make this more than a wrapper for LspFuncs, for things that shouldn't change during the execution of IdeM/IdeGhcM/IdeDeferM. Updated it to include the PID and plugins